Warn if non-trivial work is done inside .expect

darArch opened the pull request:
Added lint to warn when non-trivial work is done inside of an `.expect` call - issue [#2238](https://github.com/rust-lang-nursery/rust-clippy/issues/2338), completed as part of RustFest EU `impl Days`.

Added tests covering usage of `.expect` with `Option` and `Result` types and for expected common patterns of `.expect(&format(...))` and `.expect(format(...).as_str())`.  Per the [original issue](https://github.com/rust-lang-nursery/rust-clippy/issues/2338) the lint suggests `unwrap_or_else(|e| panic!(...))` as a replacement.

Tests are passing locally and on Travis, however, build is not succeeding overall due to a bad decrypt issue in my fork (I think this is because it is (understandably) failing to deploy the fork to Github pages due to a lack of credentials).

oli-obk commented on clippy_lints/src/methods.rs:
```diff
@@ -964,6 +992,48 @@ fn lint_or_fun_call(cx: &LateContext, expr: &hir::Expr, method_span: Span, name:
 
+/// Checks for the `EXPECT_FUN_CALL` lint.
+fn lint_expect_fun_call(cx: &LateContext, expr: &hir::Expr, method_span: Span, name: &str, args: &[hir::Expr]) {
+    #[allow(too_many_arguments)]
```
This `allow` should not be necessary anymore?

oli-obk commented on clippy_lints/src/methods.rs:
```diff
@@ -964,6 +992,48 @@ fn lint_or_fun_call(cx: &LateContext, expr: &hir::Expr, method_span: Span, name:
+            span_replace_word,
+            &format!("use of `{}` followed by a function call", name),
+            "try this",
+            format!("unwrap_or_else(|_| panic!({}))", sugg),
```
this suggestion is wrong for `Option`s (they have no closure arguments).

darArch commented:
I've made the suggested in-line changes and added tests and handling for `expect` methods implemented on custom types. The build is failing for compilation reasons which _seem_ unrelated to my changes, but if I've accidentally broken something let me know.

mati865 commented:
@darArch you can rebase on top of master branch to make CI pass.

mati865 commented:
Looks like something went wrong.
See [here](https://stackoverflow.com/questions/7929369/how-to-rebase-local-branch-with-remote-master/18442755#18442755) for instructions.

oli-obk commented on tests/ui/methods.stderr:
```diff
@@ -423,83 +423,109 @@ error: use of `unwrap_or` followed by a function call
+   --> $DIR/methods.rs:365:26
+    |
+365 |     with_none_and_format.expect(&format!("Error {}: fake error", error_code));
+    |                          ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ help: try this: `unwrap_or_else(|| panic!(&format!("Error {}: fake error", error_code)))`
```
So... I' willing to accept the PR as it is, since the suggestion is sensible for any input other than `&format!(...)`.

In the `format` case, we should get rid of the macro call and the reference and just call the `panic!` macro.

We can either merge this PR and open an issue about it, or you can fix it in this PR. What do you prefer?

darArch replied:
I see what you mean, I'll have a go at fixing it in this PR.

darArch commented:
@mati865 sorry for the confusion, I don't often do `git rebase` manually and I made a bit of a mess of it. Got there eventually.

darArch commented on clippy_lints/src/methods.rs:
```diff
@@ -964,6 +992,96 @@ fn lint_or_fun_call(cx: &LateContext, expr: &hir::Expr, method_span: Span, name:
+        let closure = if match_type(cx, self_type, &paths::OPTION) { "||" } else { "|_|" };
+        let span_replace_word = method_span.with_hi(span.hi());
+
+        if let hir::ExprAddrOf(_, ref addr_of) = arg.node {
```
@oli-obk so this big block is how I'm extracting the arguments to `format!(...)` in order to suggest passing them directly to `panic!(...)` - it's not very neat and there might be a better way of doing this but I wasn't sure what it would be or where in the codebase to look. If you could suggest a nicer way of doing this I'll happily make the change.

oli-obk replied:
I think it's fine. Macros are annoying like this. I don't think we have much other choice

darArch commented on clippy_lints/src/methods.rs:
```diff
@@ -964,6 +992,96 @@ fn lint_or_fun_call(cx: &LateContext, expr: &hir::Expr, method_span: Span, name:
+
+        if let hir::ExprAddrOf(_, ref addr_of) = arg.node {
+            if let hir::ExprCall(ref _inner_fun, ref inner_args) = addr_of.node {
+                // TODO: check if inner_fun is call to format!
```
@oli-obk also here, I wasn't sure if there was a way to easily check the function call in a similar way to how the `paths.rs` file constants can be used to check types. Again, any pointers would be appreciated.

oli-obk replied:
We do have a check for asking which macro you're in. try grepping for `"unimplemented"` or `"panic"`

darArch commented:
I tihnk I'm finished on this PR, the last set of changes should handle the `&format(...)` case suggesting `panic!(...)` directly, rather than suggesting `panic!(&format(...))`.

I also refactored some of the complexity necessary for extracting the macro arguments out of the main lint function I added in order to make it more readable. Let me know if any more changes are required or if I need to rebase again :)

oli-obk commented on clippy_lints/src/methods.rs:
````diff
@@ -329,6 +329,32 @@ declare_clippy_lint! {
+/// ```
+/// this can instead be written:
+/// ```rust
+/// foo.unwrap_or_else(|_| panic!(&format("Err {}: {}", err_code, err_msg)))
````
The docs still have the old example

darArch replied:
Bah, my bad, I knew I'd forget something. Hopefully fixed now.

oli-obk commented:
:tada: thanks! We should run this one on the rustc repo... it'll probably cause thousands of changes :laughing:

darArch commented:
@oli-obk rather you than me 😆  I think you can close https://github.com/rust-lang-nursery/rust-clippy/issues/2338 rather more safely though 😄 

Also, thanks for all the help, my first `rust-lang` or `rust-lang-nursery` project contribution 🎆 🎉

oli-obk commented:
> my first rust-lang or rust-lang-nursery project contribution

cool! great to have you on board. If you are looking for anything to do in the future, don't hesitate to ask through any communication channel
